### ISSUE TITLE
errors as data from validation failures

### ISSUE THREAD

**Issue by cainlevy** [rmosolgo/graphql-ruby#1734]
The `validate_X` methods are [recommended](http://graphql-ruby.org/mutations/mutation_authorization.html#can-this-user-modify-this-thing) to check that a specific user may mutate a specific thing in a specific way.

Sometimes this validation may fail for routine (non-malicious) reasons. For example, the resource may be locked (e.g. archived, expired). In these cases it's helpful to show a specific error to a human using [errors as data](http://graphql-ruby.org/mutations/mutation_errors#errors-as-data).

How can error data be returned from validate methods?

(Bonus: does the answer also work for prepare hooks?)

**Comment by rmosolgo** [comment]
That's a great question, I don't have an answer but I agree it should be possible 😊  

I'm currently hooking up this mutation authorization up in GitHub so it has a few other changes coming down the pike, I'll keep this in mind.

**Comment by Bahanix** [comment]
I do not know if my approach is good, but it works:

```ruby
class Resolvers::Posts::UpdateResolver < GraphQL::Function

  argument :post, !Inputs::Post::UpdateInput
  type !Types::PostFormType

  def call(obj, args, ctx)
    post = #...
    post.update(args[:post])

    OpenStruct.new({
      post: post,
      post_validation: OpenStruct.new(post.errors.to_h)
      # errors.to_h is for the example, usually that's not exactly what you want
    })
  end

end
```

```ruby
Types::PostFormType = GraphQL::ObjectType.define do
  name 'PostForm'

  field :post, Types::PostType
  field :post_validation, Types::PostValidationType
end

Types::PostValidationType = GraphQL::ObjectType.define do
  name 'PostValidation'

  field :content, types.String
end
```

The biggest downside is that every field of the OpenStruct is computed even if not requested by the client. There probably are solutions using OpenStruct and lambdas (eg: https://gist.github.com/paulfitz/10341879) but it seems even less documented.

I would be very glad having feedbacks on this pattern.

**Comment by Bahanix** [comment]
I do not know i my approach is good, but it works:

```ruby
class Resolvers::Posts::UpdateResolver < GraphQL::Function

  argument :post, !Inputs::Post::UpdateInput
  type !Types::PostFormType

  def call(obj, args, ctx)
    post = #...
    post.update(args[:post])

    OpenStruct.new({
      post: post,
      post_validation: OpenStruct.new(post.errors.to_h)
      # errors.to_h is for the example, usually that's not exactly what you want
    })
  end

end
```

```ruby
Types::PostFormType = GraphQL::ObjectType.define do
  name 'PostForm'

  field :post, Types::PostType
  field :post_validation, Types::PostValidationType
end

Types::PostValidationType = GraphQL::ObjectType.define do
  name 'PostValidation'

  field :content, types.String
end
```

The biggest downside is that every field of the OpenStruct is computed even if not requested by the client. It may become hurtful when it's not only for validations. There probably are solutions using OpenStruct and lambdas (eg: https://gist.github.com/paulfitz/10341879) but it seems even less documented.

I am very curious to have feedbacks on this.

**Comment by rmosolgo** [comment]
Fixed by #1737
